src/metalookup/core/content.py

```diff
@@ -0,0 +1,174 @@
+                )
+                text = await response.text(encoding="UTF-8")
+                return SplashResponse.parse_obj(json.loads(text))
+
+        if self._task is None:
+            self._task = asyncio.create_task(_task(), name=self.url)
+            with runtime() as t:
+                response = await self._task
+            logger.info(f"Fetched {self.url} from splash in {t():5.2f}s")
+        else:
+            logger.debug(f"Splash request already in progress, awaiting task {self._task}")
+            response = await self._task
+        return response.html, response.har
+
+    async def html(self) -> str:
+        if self._html is None:
+            self._html, self._har = await self._fetch()
+        return self._html
+
+    async def har(self) -> HAR:
+        if self._html is None:
+            self._html, self._har = await self._fetch()
+        return self._har
+
+    async def domain(self) -> str:
+        if self._domain is None:
+            host = self.tld_extractor(url=self.url.replace("http://", "").replace("https://", ""))
+            domain = [host.subdomain, host.domain, host.suffix]
+            self._domain = ".".join([element for element in domain if element != ""])
+        return self._domain
+
+    async def soup(self) -> BeautifulSoup:
+        if self._soup is None:
+            self._soup = BeautifulSoup(await self.html(), "lxml")
+        return self._soup
+
+    async def headers(self) -> dict[str, str]:
+        """A dictionary with the http headers where all keys have been converted to lowercase."""
+        # Note: See comment in self.raw_links() about the position of the await statement below.
+        har = await self.har()
+        if self._headers is None:
+            # Splash returns headers as a list of dicts with 'name' and 'value' entries for each header field.
+            # Here this list of dicts is transformed into a single dictionary with lowercase header field names.
+            # Note: As splash potentially does not normalize duplicate header fields into a comma separate list, we will
+            #       normalize them here!
+            result = dict()
+
+            # fixme: we just use the first response here which might just be a redirect...
+            #        See also https://github.com/openeduhub/metalookup/issues/85
+            for header in har.log.entries[0].response.headers:
+                lower = header.name.lower()
+                if lower not in result:
+                    result[lower] = header.value
+                else:
+                    if result[lower].endswith(";"):
+                        result[lower] = result[lower] + header.value
+                    else:
+                        result[lower] = result[lower] + ";" + header.value
+            self._headers = result
+        return self._headers
+
+    async def raw_links(self) -> list[str]:
+        # Note: The position where we await on the soup is actually critical:
```

Comment:
>  Can multiple processes both wait at the same time and start shortly after each other, so that both calculate _raw_links?
That's pretty much it!

If the suspend point (the await) is within the `if` block, then a resuming task B will continue with the `if` block (because it was suspended within it). Hence, if some other task A completes the `raw_links` computation while B was suspended in the block, B will do the same work again. If however both tasks suspend before the `if` block whichever task completes later will evaluate the `if` condition and then skip the block.

Note that this all only works, because we use single-threaded concurrency! I.e. no two tasks will ever actively run at the same time.

Reply:
Aaaah, the last paragraph resolved it for me. Thanks for clarifying. That means, in the future we have to take care not to start using multithreading.